data/templates/default.latex

```diff
@@ -362,7 +362,7 @@ $if(csl-refs)$
 \usepackage{calc}
 \newcommand{\CSLBlock}[1]{#1\hfill\break}
 \newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
-\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
+\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1\\[-1.7ex]}}
```

Comment:
Can you explain why this change fixes the problem?

Reply:
To determine where the additional vertical space was coming from I wrapped the `\parbox{}` in an `\fbox{}`. This indicated that it was from the `\break` outside the `\parbox`. By moving that inside the `\parbox` results were consistent regardless of bibitem length, but without negative adjustment were too large. By iteration arrived at the value of `-1.7ex`. I had previously been using a custom template with `\break` replaced by `\\[0.7ex]` to get spacing that matched the document. Technically why, I don't know.

Reply:
It just seems a bit "magic" and unprincipled...I wonder if there's a better way?

Reply:
What about just removing the `\break`?  Is it needed?
It seems to work fine without it and I don't get the extra space.

Reply:
During my testing I had initially tried that, but when using a more complex document using `article` and `scrartcl` the linespacing was removed between bibitems:

---

<img width="593" alt="20230905T141943-PDF Expert@2x" src="https://github.com/jgm/pandoc/assets/39587/22e00212-fb6a-4b4a-a306-9668c177f296">

---   

<img width="619" alt="20230905T141542-PDF Expert@2x" src="https://github.com/jgm/pandoc/assets/39587/54ecc924-3c39-47a1-867d-ca18534540f4">

---

but having used the above input file with `article`, `scrartcl`, `apa6` (with the `man` class option) and `beamer` it works correctly, so there must be something (not obvious) in that document which is causing the issue which I will sort out.

Reply:
Right, the problem I had experienced when only removing `\break` was caused by me using a `bibliography` with multiple references for testing. In that case the spacing between bibitems is only maintained for items that fit on one line. If in the MWE above you change the order of the two references you will see the linespacing is gobbled between the two items. The opposite of the original issue.

Reply:
Can you upload a test case? I tried to produce one myself but didn't see the problem.
<img width="687" alt="Screen Shot 2023-09-04 at 10 05 32 PM" src="https://github.com/jgm/pandoc/assets/3044/208be061-3519-4f7f-ba30-0653ba91e2cc">

Reply:
````markdown
---
fontsize: 12pt
indent: true
csl: nature
references:
- id: francois2014
  type: article-journal
  title: >-
    Lorem ipsum dolor sit amet, consectetur ullamco laboris nisi ut aliquip ex ea commodo
  container-title: Diabetologia
  author:
    - family: Williams
      given: Michael J. A.
    - family: Francois
      given: Monique E.
    - family: Baldi
      given: James C.
    - family: Manning
      given: Patrick J.
    - family: Lucas
      given: Samuel J. E.
    - family: Hawley
      given: John A.
    - family: Cotter
      given: James D.
  issued:
    date-parts:
      - - 2014
- id: WatsonCrick1953
  type: article-journal
  author:
  - family: Watson
    given: J. D.
  issued:
    date-parts:
    - - 1953
  title: Molecular aliquip ex ea commodo
  container-title: Nature
- id: zfrancois2014
  type: article-journal
  title: >-
    Lorem ipsum dolor sit amet, consectetur ullamco laboris nisi ut aliquip ex ea commodo
  container-title: Diabetologia
  author:
    - family: Zilliams
      given: Michael J. A.
    - family: Francois
      given: Monique E.
    - family: Baldi
      given: James C.
    - family: Manning
      given: Patrick J.
    - family: Lucas
      given: Samuel J. E.
    - family: Hawley
      given: John A.
    - family: Cotter
      given: James D.
  issued:
    date-parts:
      - - 2014
- id: zatsonCrick1953
  type: article-journal
  author:
  - family: Zatson
    given: J. D.
  issued:
    date-parts:
    - - 1953
  title: Molecular aliquip ex ea commodo
  container-title: Nature
nocite: |
  @*
...

Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
ullamco laboris nisi ut aliquip ex ea commodo.

Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
ullamco laboris nisi ut aliquip ex ea commodo.
````

I have increased `fontsize` to 12pt to make the difference more apparent. The difference is the same regardless of the value of `indent`, but easier to discern when `true`. It is subtle, and the font makes a difference to perception (I was using Minion Pro earlier, as that's what I use for reports, and probably because it's more condensed, makes it easier to notice). I have been taking a screenshot of the pdf output, reducing the transparency to 50% and superimposing the body text over the bibliography. The vertical space is correct after the single line bibitem, but reduced when not. I compile in a separate step with `latexmk`,  which is configured to use `xelatex`.

Reply:
OK, I can see it now (with pdflatex too).  So, in this example the vertical space between items 2 and 3 is what we want between all items?

Reply:
I notice that changing the setting of `\cslentryspacing` makes no difference whatsoever to the output.

Reply:
A bit ugly though because of the slightly larger line skip between internal lines and between entries.

Reply:
It isn't. My results, which match exactly, before I go to sleep:

```markdown
indent: false: entry-spacing="0": \setlength{\cslentryspacing}{\lineskip - \parskip}
indent: false: entry-spacing="1": \setlength{\cslentryspacing}{\lineskip}
indent: false: entry-spacing="2": \setlength{\cslentryspacing}{\lineskip + 1\baselineskip}
indent: false: entry-spacing="3": \setlength{\cslentryspacing}{\lineskip + 2\baselineskip}
indent: true: entry-spacing="0":  \setlength{\cslentryspacing}{\lineskip}
TODO: indent: true: entry-spacing="1":
TODO: indent: true: entry-spacing="2":
TODO: indent: true: entry-spacing="3":
```

I see you have closed #9032. These results are not based on that. Will look at that tomorrow, had a busy day.

Reply:
Yes, I think #9053 changes may make a difference here - I haven't had a chance to test either.

Reply:
I have tested your changes with the pandoc nightly, and with the changes from b7e1ce4 to `default.latex`. Everything is working correctly (in my opinion) when the following changes are implemented:

1.   When `entry-spacing="0"`, the correct spacing is obtained with ` 1\baselineskip - font size`. In latex the font size is held in `\f@size`. The line `\setlength{\itemsep}{#2\baselineskip}}}` accordingly needs to be able to cater for this special case. This is to comply with the [CSL specification](https://docs.citationstyles.org/en/stable/specification.html#whitespace) for `entry-spacing`:

      > Specifies vertical distance between bibliographic entries. By default (with a value of “1”), entries are separated by a single additional line-height (as set by the line-spacing attribute). Can be set to any non-negative integer to specify a multiple of this amount.

      which suggests that `0` means no *additional* spacing, not zero spacing. Also, as stated, the default is `1`, not `0`, which is currently set in `src/Text/Pandoc/Writers/LaTeX.hs:349` by `Nothing -> "0"` (can be tested with `vancouver.csl`).

2.   When using a CSL style sheet with `display="block"`, the current configuration for `CSLBlock` adds extra spacing when the block item, say `note`, is the last reference item, as in `apa-annotated-bibliography.csl`. But, if this change is implemented:

      ```patch
      - \newcommand{\CSLBlock}[1]{\hfill\break#1\hfill\break}
      + \newcommand{\CSLBlock}[1]{\hfill\break#1}
      ```

      then if the block item isn't the last item, it's not possible to compile. I'm not sure of a general solution which covers both cases. For my personal use I will use a custom template with this change.

9.   Your suggestion, which fixes the title of this PR:

   ```patch
   - \newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
   + \newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}}
   ```

Does pandoc need to consider adding `line-spacing`, or is this just assumed to match document line spacing?

Reply:
Thanks for this. I'll make these changes (except for removing the final break from CSLBlock).

I'm wondering whether, when entry spacing is set to 0, we should just use `\parskip` instead of actually using 0?  In a document that has a `\parskip` of 6pt (default for pandoc), the bibliography may look odd with 0 space between entries.  But maybe that's too opinionated a divergence.

Reply:
Currently we're ignoring `line-spacing` (though citeproc does record this).  It seems to me that nobody is really going to want a line spacing in the bibliography that differs from that of the document itself.  If people say otherwise, we could consider it.

Reply:
I've pushed some more changes; further testing welcome!

Reply:
In `default.latex`, fix typo, and use CSL name for both arguments:

```patch
- \newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
+ \newenvironment{CSLReferences}[2] % #1 hanging-indent, #2 entry-spacing
```

Reply:
Thanks, fixed.

Reply:
In `default.latex` fix rare issue that causes an extra line to be a added inside a bibitem's parbox. Occurs in the situation where the last line of a bibitem extends up to the right margin, but not quite far enough to wrap.

<img width="601" alt="20230908T092159-PDF Expert@2x" src="https://github.com/jgm/pandoc/assets/39587/7588e53e-ebd0-4ac5-b293-5d8695f98aad">

I initially never had the spaces, but added them after reading on TeX Stack Exchange (David Carlisle) that without them wrapping could be affected if the parboxes contained hard line breaks, so thought better to cater for possible future changes:(

```patch
- \newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{\strut #1 \strut}}
- \newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{\strut #1 \strut}}
+ \newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{\strut#1\strut}}
+ \newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{\strut#1\strut}}
```

Reply:
OK, done.

Reply:
> I'm wondering whether, when entry spacing is set to 0, we should just use `\parskip` instead of actually using 0? In a document that has a `\parskip` of 6pt (default for pandoc), the bibliography may look odd with 0 space between entries. But maybe that's too opinionated a divergence.

No. I think it would be incorrect in terms of the CSL spec. I believe the current implementation is accurate.

With `entry-spacing="0"` and `\setlength{\itemsep}{#2\baselineskip}}}`:

<img width="488" alt="20230908T103846-PDF Expert@2x" src="https://github.com/jgm/pandoc/assets/39587/391413f5-7ab6-4e88-9ddf-92df2ce235ca">

and `entry-spacing="0"` and `\setlength{\itemsep}{\parskip}}}`:

<img width="485" alt="20230908T103936-PDF Expert@2x" src="https://github.com/jgm/pandoc/assets/39587/76863ea6-0555-4f04-85ad-3460e8a385f2">

The spacing is approaching what it is when `entry-spacing="1"`. The paragraph spacing has nothing to do with the entry spacing, which had me confused until I read the spec more carefully

Reply:
Yes, I agree it's more accurate this way.  I just think it's a bit awkward that, with the default 6pt parskip that pandoc gives you without `indent`, you have a choice between a bigger or a smaller space between bib entries, but no way to get the same...

Reply:
I actually prefer the way it is :smiley:. 

<img width="574" alt="20230908T160345-PDF Expert@2x" src="https://github.com/jgm/pandoc/assets/39587/ac27a7a0-eb4a-4793-951f-c536b9978cfb">

This is going to be the  case with all documents where `entry-spacing="0"` and `indent: false`, regardless if using a KOMA class, the `parskip` package, or the pandoc default of `6pt plus 2pt minus 1pt`. In my custom template I used to have the `entry-spacing` depend on `indent`. But, I was wrong, due to my ignorance from not having read the spec, assuming 0 meant zero spacing (not zero *additional* spacing), 1 was line spacing (actually 0)... and which clearly others do too e.g. #7296. But, I can appreciate the visual appeal:

<img width="559" alt="20230908T160629-PDF Expert@2x" src="https://github.com/jgm/pandoc/assets/39587/cf8684af-bb33-4b1a-b234-6ee91d000f66">

rather than moving up to `entry-spacing="1"`:

<img width="558" alt="20230908T160659-PDF Expert@2x" src="https://github.com/jgm/pandoc/assets/39587/0ba385c6-5171-440e-86f0-9f2a3bf8347f">

However, if you decide to implement this please don't make it the default, but rather an option, or better yet as a filter, so that at least the default is compliant with the spec. Of course, for those wanting something special, it is always possible to use `header-includes`:

```latex
\usepackage{etoolbox}
\AtBeginEnvironment{CSLReferences}[\setlength{\itemsep}{ANYTHING YOU WISH}
```

I have completed testing the latex CSL functionality. I have tested with the KOMA, APA, and standard classes, together with beamer. All combinations of linespacing (single, onehalf and double) work correctly with `entry-spacing` values of 0, 1 and 2, using bibliographies of approx 50–200 items. Thanks very much for your efforts with this, I am really happy with the results.

Reply:
Thanks very much for the extensive testing!  I'll close this.